src/main/java/bio/terra/service/filedata/FileMetadataUtils.java

```diff
@@ -133,4 +136,48 @@ public static List<String> extractDirectoryPaths(String path) {
     }
     return allPaths;
   }
+
+  /**
+   * Results of the successful md5 comparison of user a specified MD5 against a cloud provided MD5
+   *
+   * @param effectiveMd5 The final md5 to record
+   * @param type Which md5 was actually used
+   */
+  public record Md5ValidationResult(String effectiveMd5, Md5Type type) {
+    public enum Md5Type {
+      USER_PROVIDED, // When the user md5 is used (e.g. when it's not present in the cloud)
+      CLOUD_PROVIDED, // When the cloud md5 is used
+      NEITHER // When neither value is provided
+    }
+  }
+
+  /**
+   * Validates that a user specified MD5 checksum matches with a cloud provided MD5 checksum. If the
+   * user provided checksum is null or empty, then assume the cloud checksum is valid and return
+   * that
+   *
+   * @param userSpecifiedMd5 A hex representation of the user specified MD5 file checksum
+   * @param cloudMd5 A hex representation of the MD5 file checksum for the file object
+   * @param sourcePath The cloud path where the file lives
+   * @return A hex representation of the file's md5 hash
+   * @throws InvalidFileChecksumException if the specified checksums don't match
+   */
+  public static Md5ValidationResult validateFileMd5ForIngest(
```

Comment:
Looking at the logic in this method as well as how it's called, I wonder about making at least `userSpecifiedMd5` an `Optional<String>` instead?  That's where my mind goes when I see emptiness checks.

Reply:
Yeah, I agree though there's a codebase consistency issue where we don't really use optional parameters anywhere else